### PR TITLE
Partially implemented and tested SchedulignService

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

Flesh out internal design of implementation of `SchedulingService`. Implemented a portion of the logic and added a portion of test for the implementation. More logic and tests will be added later, but I want to get some feedback on a partially working implementation

## Testing
Some unit tests were added that leverage new added or tweaked fakes.

### CODE REVIEW

**Comment by bidetofevil** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1426?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1426** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1426?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1425** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1425?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1384** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1384?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/worker/Worker.kt
Comment (bidetofevil) [inline embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/worker/Worker.kt:60]: This is the worker that runs jobs that queue up the network request send and blocks on execution, waiting for a response.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/SupportedEnvelopeType.kt
Comment (bidetofevil) [inline embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/SupportedEnvelopeType.kt:10]: Need to know which endpoint an envelope type can go to to implement the logic to block sending to specific endpoints if the server returns a 429.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingService.kt
Comment (bidetofevil) [inline embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingService.kt:17]: Using this in the DeliveryModule for now. If this is not useful in the future when the implementation is fully rolled out, we can delete it then.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [inline embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:20]: The general logic is as follows:

- When `onPayloadIntake()` is invoked, schedule a job on the `scheduleWorker` (using the non-IO background worker) to loop through the available payloads to be sent and attempts to send them if we aren't already running such a loop

- The loop will query for all the available payloads, and as long as a payload is eligible to be sent, queue it up to be sent

- When the loop finishes, it checks to see if there are additional payloads to send since the last query. Any payload that's in progress will be skipped (the current implementation can probably be optimized to do better filtering). The loop ends if it doesn't find any new payloads, and if there are payloads that ought to be retried, the earliest time will be used for a job to be scheduled to run then to check for payloads to deliver. The loop, at that point, should find at least that one payload and try to send it.

- When a payload is sent successfully, it is deleted via the storage service. 

- If it's not successful, it will be added to a Map of payloads that will be retried, with some metadata about when it should next be attempted, and how many failures there had been. The latter will be used to put a cap on how many retries are done (to be implemented). If a payload send retry is successful, this payload is removed. If it fails again, the metadata is updated.

- To facilitate this, `PayloadReference` was added as an abstraction of the cached payload file. A concept like this is needed so we don't just use the `File` reference, which is unstable and could be duplicated. The attributes in this data class should uniquely identify a payload on disk to be sent, and also provide enough metadata required for various parts of the delivery process.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (fractalwrench) [inline embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:36]: One trade-off of this approach is that if the `storageService` is saturated with less important payloads such as logs, any new important payloads such as crashes/sessions might not get scheduled for a long time. In extreme cases where the cache limit is reached, that will lead to data loss of important payloads rather than less important ones.

IMO the `SchedulingService` should probably recheck what files are available whenever `onPayloadIntake` is called to avoid this scenario, given that the aim of this project is to avoid all avoidable sources of data loss. This doesn't necessarily have to mean disk writes each time if the `StorageService` maintains an internal list that is updated whenever a file is added/deleted.

**Review by fractalwrench — APPROVED** [review]
Approving so we can merge & iterate given this isn't hooked up to anything in production yet.

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/worker/Worker.kt
Comment (bidetofevil) [reply embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/worker/Worker.kt:60]: Ooh good call. Will add a TODO

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/RequestExecutionService.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/RequestExecutionService.kt:16]: Is it possible for the lambda to be non-null but the stream it generates is null? I'm open to handling the null case before calling if that's the case.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:30]: Gah. Twas a mistake. I only need threadsafety. Good catch. Will modify.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:36]: Hmm. Yeah, that's good to support. I think I'll do something a bit different - every call where there's an active loop, I'll notify it to requery for additional files, basically resetting the loop. This might even allow me to remove the additional check after the loop finishes to pick up newly inserted payloads.

Let me do that!

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:68]: So the initial thought was that `filter` will do the filtering at the beginning, but one condition for ineligibility is retry time - by the time it gets to a payload, it may be eligible for delivery. 

That said, if we keep the additional check at the end of the loop, any payloads that became eligible will be picked up then, so that will probably work just as well. Yeah, let me do that.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:101]: I expect `PayloadStorageService` to handle any deletion failures

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:106]: Haha fair. Will make it less dense

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt:124]: Yeah, it is a bit of an awkward fit. I can pull the logic of the `onPayloadIntake()` into a private function and have that called here instead, as you suggested.

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/PayloadReference.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/PayloadReference.kt:6]: Perfect! Will do

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt:5]: Definitely combine, as we discussed

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt:6]: In one of the doc's comments from Fredric, he suggested that the priority is inherent in the file name's format if it's sorted alphabetically, e.g. `p1_<timestamp>_type_v1.json`. How do you feel about that?

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt
Comment (bidetofevil) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt:7]: Totally fine. I am treating this as a blob that I'll pass around without knowing anything about, so that works fine for me.

File: embrace-test-fakes/src/main/kotlin/io/embrace/android/embracesdk/fixtures/DeliveryFixtures.kt
Comment (bidetofevil) [reply embrace-test-fakes/src/main/kotlin/io/embrace/android/embracesdk/fixtures/DeliveryFixtures.kt:8]: nice

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/RequestExecutionService.kt
Comment (fractalwrench) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/RequestExecutionService.kt:16]: Yep that's possible - we can always return a lambda, but the inputstream might fail to open/error for whatever reason

File: embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt
Comment (fractalwrench) [reply embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/storage/StorageService2.kt:6]: That sounds reasonable. We do include UUIDs in the filename right now so might want to move that around.

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt
Comment (bidetofevil) [inline embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt:53]: No a big fan of the current abstraction. I don't want to change things for the old code, but I think the new delivery layer should differentiate between when a request has been sent and when it hasn't and have values that aren't really instances of `ApiResponse`

**Review by bidetofevil — COMMENTED** [review]
Did everything in the review comments except for merging StorageService2 and PayloadStorageService. Will do that in a subsequent PR.

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt
Comment (fractalwrench) [reply embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt:53]: Would renaming `ApiResponse` make this clearer? What other values would be needed by the SDK & how would behavior differ?

**Review by fractalwrench — APPROVED** [review]
LGTM

File: embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt
Comment (bidetofevil) [reply embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/api/ApiResponse.kt:53]: Renaming would help. Basically the grouping is:

- Request sent
  - Response received
    - <various different classifications of responses>  
  - No response
    - Client error
    - Server error
    - Timeout
- Request not sent
  - No payload
  - Request not eligible (e.g. 429'd)
  - SDK/environment not ready (e.g. no network)
  - Client error

All of these will potentially affect when a payload is retried, and any clean up or error reporting we need to do. Renaming would work if we carefully manage all the places that check this, and deal with any fall-through states or explicitly handle new ones.

Part of me wonders if a new start is less risky.... but maybe I'll try that first and back off if it gets too hairy.

**Comment by bidetofevil** [comment]
### Merge activity

* **Sep 27, 2:52 PM EDT**: @bidetofevil started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1426).

**Pull request merged**
